File: 1520-maximum-number-of-non-overlapping-substrings/1520-maximum-number-of-non-overlapping-substrings.java
class Solution {
    public List<String> maxNumOfSubstrings(String s) {
        int n = s.length();
        int[] first = new int[26];
        int[] last = new int[26];
        Arrays.fill(first, Integer.MAX_VALUE);

        for (int i = 0; i < n; i++) {
            int c = s.charAt(i) - 'a';
            first[c] = Math.min(first[c], i);
            last[c] = i;
        }

        List<int[]> intervals = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            if (i != first[s.charAt(i) - 'a']) continue;

            int start = i;
            int end = last[s.charAt(i) - 'a'];
            boolean valid = true;

            for (int j = start; j <= end; j++) {
                int c = s.charAt(j) - 'a';
                if (first[c] < start) {
                    valid = false;
                    break;
                }
                end = Math.max(end, last[c]);
            }

            if (valid) {
                intervals.add(new int[]{start, end});
            }
        }

        // Greedy: sort by end ascending, pick non-overlapping
        intervals.sort((a, b) -> a[1] - b[1]);

        List<String> result = new ArrayList<>();
        int lastEnd = -1;

        for (int[] interval : intervals) {
            if (interval[0] > lastEnd) {
                result.add(s.substring(interval[0], interval[1] + 1));
                lastEnd = interval[1];
            }
        }

        return result;
    }
}

File: 1520-maximum-number-of-non-overlapping-substrings/1520-maximum-number-of-non-overlapping-substrings.cs
public class Solution {
    public IList<string> MaxNumOfSubstrings(string s) {
        int n = s.Length;

        // Step 1: first and last occurrence of each letter
        var first = new int[26];
        var last = new int[26];
        Array.Fill(first, int.MaxValue);
        Array.Fill(last, -1);

        for (int i = 0; i < n; i++) {
            int c = s[i] - 'a';
            first[c] = Math.Min(first[c], i);
            last[c] = i;
        }

        // Step 2: For each index, find the smallest valid window starting there
        var intervals = new List<(int start, int end)>();

        for (int i = 0; i < n; i++) {
            if (i != first[s[i] - 'a']) continue;  // must start at first occurrence

            int end = last[s[i] - 'a'];
            for (int j = i; j <= end; j++) {
                int c = s[j] - 'a';
                if (first[c] < i) { end = -1; break; }  // a needed letter starts before i → invalid
                end = Math.Max(end, last[c]);           // expand window
            }

            if (end != -1)
                intervals.Add((i, end));
        }

        // Step 3: Greedy interval scheduling — sort by end, pick non-overlapping
        intervals.Sort((a, b) => a.end.CompareTo(b.end));

        var result = new List<string>();
        int lastEnd = -1;
        foreach (var (start, end) in intervals) {
            if (start > lastEnd) {
                result.Add(s.Substring(start, end - start + 1));
                lastEnd = end;
            }
        }

        return result;
    }
}